packages/storage/amplify_storage_s3_dart/test/storage_s3_service/transfer/database_html_test.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+@TestOn('browser')
+import 'dart:html';
+
+import 'package:amplify_storage_s3_dart/src/storage_s3_service/transfer/database/database_html.dart';
+import 'package:amplify_storage_s3_dart/src/storage_s3_service/transfer/database/transfer_record.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('TransferDatabase for web', () {
```

Comment:
These tests make the code a bit brittle since they test implementation and cannot be run out of order (are inter-dependent). I would recommend some functional testing, i.e. tests which can run on Web or VM and only test the exposed API of `TransferDatabase`, and making tests self-contained so they are not dependent on any test running before/after it.

Reply:
tests can be run out of order and have not dependency on each other. do they fail for you?

Reply:
Gotcha - I was doing something wrong. The other point still holds, though. We are testing implementation which means the impl cannot easily change.

Reply:
in terms of testing functionality and not implementation I beleive all tests here test for functionality exept the part at line 37 which checks `expect(recordId.startsWith(localStorageKeyPrefix), isTrue)`. I agree it is testing implementation and will clean up. I wanted to make sure this is the one that you referring to.

Reply:
`localStorage` is an implementation detail. Functional tests should allow swapping out implementation with IndexedDB or whatever without changes to the tests.

Reply:
we initially mocked the storage and then changed to use window.localStorage, https://github.com/aws-amplify/amplify-flutter/pull/2631#discussion_r1096107807

Reply:
Yeah, sorry for the confusion... this is fine for now. Overall, what I mean is that with a test like this:

```dart
test('insertTransferRecord to local storage', () async {
  final recordId =
      await transferDatabase.insertTransferRecord(testTransferRecord);
  final actual = window.localStorage[recordId];
  expect(actual, testTransferRecordJsonString);
});
```

It's tied to the implementation. So in order to swap out `localStorage` for something else inside the database, you would need to update this test, which introduces a maintenance burden.

Compare that to the rest of the tests which do not reference `localStorage` but work solely off the public interface. With just those tests, I think you get the same level of assurance in terms of behavior, but you are free to change the implementation at any point. And then also these tests could be reused for the IO database, again reducing the maintenance burden of two parallel sets of tests.

I will leave it up to you - these are just my two cents.

Reply:
updated test to not use localStorage